### PR TITLE
Feat: add support for Conv2D DoRA

### PR DESCRIPTION
**Pull request by sayakpaul**
TODO

- [x] Incorporate `_apply_dora()` in the `forward()` of Conv2D
- [x] `merge()`
- [x] `unmerge()`
- [x] Tests
- [x] Docs

### CODE REVIEW

**Comment by sayakpaul** [comment]
@BenjaminBossan would love to get your thoughts on the changes I have introduced so far here.

**Comment by BenjaminBossan** [comment]
Thanks a lot Sayak, this already looks pretty good. Probably the main thing to get right is the dimensions across which to normalize, which does look correct to me. Maybe there is also some room to speed up the application of the norm, not sure.

As a next step, let's add a test case for Conv2d + DoRA. This should be as easy as adding an example or two here:

https://github.com/huggingface/peft/blob/7e84dec20b3106bdd0a90ba8e80187f0aec835b7/tests/test_custom_models.py#L58-L70

Let's do this and check if the tests pass, then plan the next steps.

**Comment by sayakpaul** [comment]
@BenjaminBossan I have added tests. However, 4 tests related to "merging" are currently failing. I would appreciate another round of review; maybe something pesky will come out that's making those tests fail. LMK.

**Comment by sayakpaul** [comment]
@BenjaminBossan I think this is ready for a review now.

**Comment by sayakpaul** [comment]
Cc: @nbasyl if you want to give this a look as well.

File: src/peft/tuners/lora/layer.py [hunk 1]
```diff
@@ -797,6 +838,38 @@ def get_delta_weight(self, adapter) -> torch.Tensor:
 
         return output_tensor
 
+    def _get_weight_norm(self, weight, lora_weight, scaling) -> torch.Tensor:
+        # calculate L2 norm of weight matrix, channel-wise
+        weight = weight + scaling * lora_weight
+        weight_norm = weight.norm(p=2, dim=(1, 2, 3))
+        weight_norm = weight_norm.reshape((1, self.get_base_layer().out_channels, 1, 1))
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/layer.py:845]: Would this work instead:

`weight_norm = weight.norm(p=2, dim=(1, 2, 3), keepdim=True).transpose(1, 0)`

File: src/peft/tuners/lora/layer.py [hunk 2]
```diff
@@ -797,6 +838,38 @@ def get_delta_weight(self, adapter) -> torch.Tensor:
...
+        # see section 4.3 of DoRA (https://arxiv.org/abs/2402.09353)
+        # "[...] we suggest treating ||V +∆V ||_c in
+        # Eq. (5) as a constant, thereby detaching it from the gradient
+        # graph. This means that while ||V + ∆V ||_c dynamically
+        # reflects the updates of ∆V , it won’t receive any gradient
+        # during backpropagation"
+        weight_norm = weight_norm.detach()
+        mag_norm_scale = magnitude / weight_norm
+        result_dora = (mag_norm_scale - 1) * (
+            F.conv2d(x, weight, None, base_layer.stride, base_layer.padding, base_layer.dilation, base_layer.groups)
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/layer.py:868]: Can we pass all args after `weight` as keyword args please?

File: tests/test_custom_models.py [hunk 3]
```diff
@@ -646,7 +648,10 @@ def test_parameters_after_loading_model(self, test_name, model_id, config_cls, c
     @parameterized.expand(TEST_CASES)
     def test_disable_adapters(self, test_name, model_id, config_cls, config_kwargs):
         X = self.prepare_inputs_for_testing()
+        print(f"Class name: {self.transformers_class.__name__}")
         model = self.transformers_class.from_pretrained(model_id).to(self.torch_device).eval()
+        for k in X:
+            print(k, X[k].shape)
```
Comment (BenjaminBossan) [inline tests/test_custom_models.py:654]: Please remove.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
This looks pretty good, thanks a lot for enabling DoRA with Conv2d. Curious to see if we can get nice improvements for diffusion models.

Implementation-wise, I have only a few minor comments. On top of those, could you please document the new changes here:

- The docstring of `use_dora` inside of `tuners/lora/config.py` still mentions that only linear layers are supported.
- Same for the help of `use_dora` in the same file.
- `docs/source/developer_guide/lora.md` also needs updating with regard to the supported layers.

Before merging, it would be nice to have another review.

Edit: Please also merge with recent main, as there could be merge conflicts with #1518 otherwise.

File: src/peft/tuners/lora/layer.py (hunk 1 above)
Comment (sayakpaul) [reply src/peft/tuners/lora/layer.py:845]: Could work but I think the code is slightly more readable in the current form.

**Comment by sayakpaul** [comment]
@BenjaminBossan resolved your comments. Thank you!

File: src/peft/tuners/lora/layer.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/peft/tuners/lora/layer.py:852]: We can leave the reshape (though view is mostly used otherwise), but I think we should avoid the call to `self.get_base_layer().out_channels`, as we have all the information we need without it. Maybe a comment that documents that shapes could be added instead.

Comment (sayakpaul) [reply src/peft/tuners/lora/layer.py]: Fixed in 641c5b5.

File: src/peft/tuners/lora/config.py [hunk 4]
```diff
@@ -104,7 +104,10 @@ class LoraConfig(PeftConfig):
         use_dora (`bool`):
             Enable 'Weight-Decomposed Low-Rank Adaptation' (DoRA). This technique decomposes the updates of the weights
             into two parts, magnitude and direction. Direction is handled by normal LoRA, whereas the magnitude is
-            handled by a separate learnable parameter. This can improve the performance of LoRA, especially at low
+            handled by a separate learnable parameter. This can improve the performance of LoRA especially at low
+            ranks. Right now, DoRA only supports linear and Conv2D layers. DoRA introduces a bigger overhead than pure
```
Comment (BenjaminBossan) [inline src/peft/tuners/lora/config.py:108]: Oops, I must have broken this with a merge resolution. Thanks for fixing!

**Review by BenjaminBossan — APPROVED** [review]
Thanks a lot, Sayak, the PR looks great. Let's see if Shih-yang also has time to take a look. Otherwise, I'd be glad to have a second pair of eyes by @younesbelkada or @pacman100.

**Comment by BenjaminBossan** [comment]
Thanks so much Sayak.
